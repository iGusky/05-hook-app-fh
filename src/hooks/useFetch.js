import { useEffect, useRef, useState } from 'react'

export const useFetch = ( url ) => {
  
  // Se crea una referencia para saber si el componente está montado.
  const isMounted = useRef(true);
  const [state, setState] = useState({ data: null, loading: true, error: null});

  //Cuando el componente se desmonta (return) el useRef se vuelve falso.
  useEffect(() => {
    return () => {
      isMounted.current = false;
    }
  }, [])

  useEffect( () => {

    setState({ data: null, loading: true, error: null})

    fetch( url )
      .then( response => response.json())
      .then( data => {
        
        //Si es falso no se cambiará el estado.
        if(isMounted.current) { 
          setState({
            loading: false,
            error: null,
            data
          });
        }

      })
      .catch( () => {
          setState({
              data: null,
              loading: false,
              error: 'No se pudo cargar la data'
          });
      })
  }, [ url ]);

  return state;

}
