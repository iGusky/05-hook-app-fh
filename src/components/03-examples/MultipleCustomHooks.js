import React from 'react';
import useCounter from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch';
import './effects.css';

export const MultipleCustomHooks = () => {

  //Usaremos el hook useCounter (custom).
  const { counter, increment} = useCounter(1);
  //Se envia el url al hook useFetch (custom).
  const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
  //Se valida si viene la nada, si es verdadero se extrae la informaci√≥n.
  const { author, quote } = !!data && data[0];

  
  return (
    <>
      <h1>BreakingBad Quotes</h1>
      <hr />

      {
        // Uso de operador terniario...
        loading ? 
          (
            <div className="alert alert-info text-center">
              Loading...
            </div>
          )
        :
          (
            <>
            <blockquote className="blockquote text-right">
              <p className="mb-0">{ quote }</p>
              <footer className="blockquote-footer"> { author }</footer>
            </blockquote>

              <button 
                className = "btn, btn-primary"
                onClick= { () => increment() }
              >
                Siguiente frase
              </button>
            </>
          )
      }

   

    </>
  )
}
