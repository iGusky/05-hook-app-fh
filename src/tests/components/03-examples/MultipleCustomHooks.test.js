import { shallow } from 'enzyme' 
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks'
import { useFetch } from '../../../hooks/useFetch'
jest.mock('../../../hooks/useFetch');

describe('Pruebas en <MultipleCustomHooks />', () => {
    
    test('debe mostrarse correctamente', () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow( <MultipleCustomHooks /> )
        expect( wrapper ).toMatchSnapshot();
    });

    test('debe retornar informaciÃ³n de fetch', () => {
        
        useFetch.mockReturnValue({
            data: [{
                author: 'Gustavo',
                quote: 'Hola mundo'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        
        expect( wrapper.find('.alert').exists() ).toBe(false);
        expect( wrapper.find('.mb-0').text().trim() ).toBe('Hola mundo');
        expect( wrapper.find('footer').text().trim() ).toBe('Gustavo');
        
    });
    
    

})
