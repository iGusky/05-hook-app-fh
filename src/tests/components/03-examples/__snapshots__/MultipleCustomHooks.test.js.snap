// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> debe mostrarse correctamente 1`] = `
<Fragment>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
</Fragment>
`;
